flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1860,18 +1860,18 @@ fir::ExtendedValue
 IntrinsicLibrary::genAssociated(mlir::Type resultType,
                                 llvm::ArrayRef<fir::ExtendedValue> args) {
   assert(args.size() == 2);
-  // TODO: Runtime would be better to deal with all the cases with TARGET
-  // arguments.
-  if (!isAbsent(args[1]))
-    TODO(loc, "ASSOCIATED intrinsic with TARGET argument");
-  return args[0].match(
-      [&](const fir::MutableBoxValue &x) -> fir::ExtendedValue {
-        return Fortran::lower::genIsAllocatedOrAssociatedTest(builder, loc, x);
-      },
-      [&](const auto &) -> fir::ExtendedValue {
-        fir::emitFatalError(loc,
-                            "associated arg not lowered to MutableBoxValue");
-      });
+  auto *pointer =
+      args[0].match([&](const fir::MutableBoxValue &x) { return &x; },
+                    [&](const auto &) -> const fir::MutableBoxValue * {
+                      fir::emitFatalError(loc, "pointer not a MutableBoxValue");
+                    });
+  if (isAbsent(args[1]))
+    return Fortran::lower::genIsAllocatedOrAssociatedTest(builder, loc,
+                                                          *pointer);
+  auto pointerBox = builder.createBox(
+      loc, genMutableBoxRead(builder, loc, *pointer).getAddr());
```

Comment:
The canonical way to get a fir.box from a `MutableBox` is to do get the related `fir.ref<fir.box>` with `Fortran::lower::getMutableIRBox(builder, loc, *pointer)`, and then you can load that (you can get the box type for the LoadOp with pointer->getBoxTy()).

`getMutableIRBox` deals with the cases where the mutable box is not a descriptor but a set of variables. The code you wrote leads to a similar behavior in most cases, with the exception that it may create redundant boxes in a few cases where `genMutableBoxRead` gets rid of the descriptor because it knows the pointer is contiguous.

Reply:
To be more precise, I think you can replace:
```
  auto pointerBox = builder.createBox(
      loc, genMutableBoxRead(builder, loc, *pointer).getAddr());
```

By:

```
auto pointerBoxRef = Fortran::lower::getMutableIRBox(builder, loc, *pointer);
auto pointerBox = builder.create<fir::LoadOp>(loc, pointerBoxRef);
```

Then, you'll see that it will actually simplify the IR in your test from:

```
  ! CHECK: %[[scalar:.*]] = fir.load %[[arg0]] : !fir.ref<!fir.box<!fir.ptr<f32>>>
  ! CHECK: %[[addr2:.*]] = fir.box_addr %[[scalar]] : (!fir.box<!fir.ptr<f32>>) -> !fir.ptr<f32>
  ! CHECK: %[[sbox:.*]] = fir.embox %[[addr2]] : (!fir.ptr<f32>) -> !fir.box<f32>
```

into

```
 ! CHECK: %[[sbox:.*]] = fir.load %[[arg0]] : !fir.ref<!fir.box<!fir.ptr<f32>>>
 ! CHECK: %[[sbox1:.*]] = fir.convert %[[sbox]] : (!fir.box<!fir.ptr<f32>>) -> !fir.box<none>
```

Reply:
Would it be possible in a separate PR to rename `getMutableIRBox` to something like `genLoadMutableBox`? `get` seems a bit confusing since getters are typically retrieving information from data structures, while generate (`gen`) is typically used to suggest some chunk of IR is going to be created.

Reply:
`getMutableIRBox` does not load the fir.ref<fir.box>. It is a no-op if the allocatable/pointer is primarily tracked as a Fortran descriptor. In case the allocatable/pointer is tracked by a set of local variables, it generates code to sync the Fortran descriptor with their value before returning the fir.ref<fir.box>.
So `genLoadMutableBox` sounds misleading to me. What about renaming it to `syncMutableIRBox` ?

Reply:
I've updated the code; thanks for the detailed suggestion.